### PR TITLE
Added table, index, and view caching for performance

### PR DESCRIPTION
**Pull request by zachmu**

### CODE REVIEW

**Comment by zachmu** [comment]
#benchmark

File: go/libraries/doltcore/sqle/dsess/session_cache.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/doltdb"
+)
+
+// SessionCache caches various pieces of expensive to compute information to speed up future lookups in the session.
+// No methods are thread safe.
+type SessionCache interface {
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/dsess/session_cache.go:27]: do we need an interface for this? we don't have any polymorphism here

File: go/libraries/doltcore/sqle/tables.go [hunk 2]
```diff
@@ -248,7 +248,12 @@ func (t *DoltTable) DataCacheKey(ctx *sql.Context) (doltdb.DataCacheKey, bool, e
 	if err != nil {
 		return doltdb.DataCacheKey{}, false, err
 	}
-	return doltdb.NewDataCacheKey(r), true, nil
+	key, err := doltdb.NewDataCacheKey(r)
+	if err != nil {
+		return doltdb.DataCacheKey{}, false, err
+	}
+
+	return key, true, nil
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/tables.go:256]: we do we have a boolean signal here if we're not using it? future use?

File: go/libraries/doltcore/sqle/tables.go [hunk 3]
```diff
@@ -267,12 +272,46 @@ func (t *DoltTable) getRoot(ctx *sql.Context) (*doltdb.RootValue, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	if !ok {
+		tbl, err := t.DoltTable(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return index.DoltIndexesFromTable(ctx, t.db.Name(), t.tableName, tbl)
```
Comment (andrew-wm-arthur) [inline go/libraries/doltcore/sqle/tables.go:285]: we don't populate the cache here, but we also never his this block

**Review by andrew-wm-arthur — APPROVED** [review]
LGTM, couple comments

**Comment by zachmu** [comment]
#benchmark

File: go/libraries/doltcore/sqle/tables.go (hunk 2 above)
Comment (zachmu) [reply go/libraries/doltcore/sqle/tables.go:256]: This implementation always returns true, but there are others that only return false

File: go/libraries/doltcore/sqle/tables.go (hunk 3 above)
Comment (zachmu) [reply go/libraries/doltcore/sqle/tables.go:285]: Adding a comment to clarify

**Pull request merged**